scanspec/service.py

```diff
@@ -1,19 +1,148 @@
+    """
+
+    axis: str
+    """A fixed reference that can be scanned. i.e. a motor, time or
+    number of repetitions.
+    """
+    lower: Optional[Points]
+    """The lower bounds of each midpoint (used when fly scanning)"""
+    midpoints: Optional[Points]
+    """The centre points of the scan"""
+    upper: Optional[Points]
+    """The upper bounds of each midpoint (used when fly scanning)"""
+
+
+@dataclass
+class PointsRequest:
+    """ The highest level of the getPoints query, allowing users to customise their
+    return data from the points present in the scan to some metadata about them
+    """
+
+    axes: List[AxisFrames]
+    total_frames: int
+    returned_frames: int
+
+
+# Chacks that the spec will produce a valid scan
 def validate_spec(spec: Spec) -> Any:
+    """ A query used to confirm whether or not the Spec will produce a viable scan"""
     # apischema will do all the validation for us
     return spec.serialize()
 
 
-schema = graphql_schema(query=[validate_spec])
+# Returns a full list of points for each axis in the scan
+# TODO Update max_frames with a more sophisticated method of reducing scan points
+def get_points(spec: Spec, max_frames: Optional[int] = 200000) -> PointsRequest:
+    """A query that takes a Spec and calculates the points present in the scan
+    (for each axis) plus some metadata about the points.
+
+    Arguments:
+            [spec]: [The specification of the scan]
+            [max_frames]: [The maximum number of frames the user wishes to receive]
+
+    Returns:
+        [PointsRequest]: [A dataclass containing information about the scan points
+                            present in the spec]
+    """
+    dims = spec.create_dimensions()  # Grab dimensions from spec
+    path = Path(dims)  # Convert to a path
+    total_frames = len(path)  # Capture the total length of the path
+
+    # Limit the consumed data to the max_frames argument
+    # # WARNING: path object is consumed after this statement
+    if max_frames is None:
+        # Return as many frames as possible
+        returned_frames = len(path)
+        chunk = path.consume(len(path))
+
+    elif max_frames >= len(path):
+        # Return all of the frames within that selection
+        returned_frames = len(path)
+        chunk = path.consume(len(path))
+
+    else:
+        # Cap the frames by the max limit
+        returned_frames = max_frames
+        chunk = path.consume(max_frames)
+
+    # POINTS
+    scan_points = [
+        AxisFrames(
+            axis,
+            Points(chunk.lower.get(axis)),
+            Points(chunk.midpoints.get(axis)),
+            Points(chunk.upper.get(axis)),
+        )
+        for axis in spec.axes()
+    ]
+
+    return PointsRequest(scan_points, total_frames, returned_frames)
+
+
+# Define the schema
+schema = graphql_schema(query=[validate_spec, get_points])
+
+
+def schema_text() -> str:
+    return graphql.utilities.print_schema(schema)
 
 
 def run_app(cors=False):
```

Comment:
For compatibility with the GDA ADE, the service needs to be able to run on a port other than 8080.
You can get a custom port as an argument in cli.py and pass it to this function. It then needs to be passed to the inner `run_app` function e.g.
```python
web.run_app(app, port=port)
```

Reply:
Added in this functionality. Can be called with scanspec service --port <port number>